### DISCUSSION TITLE
Train Embedding Layer with LoRA

### DISCUSSION THREAD

**Discussion by RADxEEn** [huggingface/peft#1441] [General]
Hey, so I know there was a similar question like it. But I am facing a problem
So, I not only want to update the layer, but also update the embedding layer. So for it I use 
modules_to_save=["embed_tokens"]

but after training there should be weight change between base model embed_tokens_weight and modules_to_save Embedding weight. But when I check on their mean this is 0, that means it didn't update my embedding layer weight but, it updated the layers weight. So, I guess somehow my embedding layer is not connecting or participating in training. 
`merged_model_f_n.base_model.model.model.lm_head.original_moduleSS.weight-merged_model_f_n.base_model.model.model.lm_head.modules_to_save.default.weight`

```
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0', dtype=torch.float16,
       grad_fn=<SubBackward0>)
```

```
PeftModel(
  (base_model): LoraModel(
    (model): PeftModelForCausalLM(
      (base_model): LoraModel(
        (model): MistralForCausalLM(
          (model): MistralModel(
            (embed_tokens): ModulesToSaveWrapper(
              (original_module): Embedding(67840, 4096)
              (modules_to_save): ModuleDict(
                (default): Embedding(67840, 4096)
              )
            )
```

Will be happy to accept any suggestion. Thanks

**Comment by BenjaminBossan** [comment]
This is quite strange and I could not reproduce the issue. Your suggestion to add `modules_to_save=["embed_tokens"]` is correct. However, this should create a copy of the original weights, so they should not be all zeros. Could you please provide a complete code example to reproduce the issue?

**Reply by RADxEEn** [reply]
Sure here the glimpse of it

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)
model = AutoModelForCausalLM.from_pretrained(model_id, quantization_config=bnb_config, device_map="auto")

tokenizer = AutoTokenizer.from_pretrained(model_id, token = token, add_eos_token=True)

from peft import prepare_model_for_kbit_training

model.gradient_checkpointing_enable()
model = prepare_model_for_kbit_training(model)
param = model.model.embed_tokens.weight
param.data = param.data.to(torch.float32)

def print_trainable_parameters(model):
    """
    Prints the number of trainable parameters in the model.
    """
    trainable_params = 0
    all_param = 0
    for _, param in model.named_parameters():
        all_param += param.numel()
        if param.requires_grad:
            trainable_params += param.numel()
    print(
        f"trainable params: {trainable_params} || all params: {all_param} || trainable%: {100 * trainable_params / all_param}"
    )

from peft import LoraConfig, get_peft_model, PeftModel

lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj","o_proj", "gate_proj","down_proj","up_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    modules_to_save=["embed_tokens", "lm_head"]
)

peft_model = get_peft_model(model, lora_config)
# peft_model = PeftModel.from_pretrained(model, "/content/llm_bang_mistral/checkpoint-40", is_trainable=True)
print_trainable_parameters(peft_model)

model.add_adapter(lora_config, adapter_name="bn_adapter")
model_args=(...some_training_arguments/hyperparameter)
trainer = SFTTrainer(
    model=model,
    train_dataset=train_data,
    eval_dataset=eval_data,
    peft_config=lora_config,
    dataset_text_field="merged",
    # max_seq_length=2048,
    tokenizer=tokenizer,
    packing=False,
    args=model_args,
)
```

**Reply by BenjaminBossan** [reply]
At what point exactly do you find the embeddings weights to be 0? Before calling `trainer.train()`, after, or both?

**Reply by RADxEEn** [reply]
So, after training, I got the model with adapter checkpoint right. So, I tested by taking the difference between the base model and the merged adapter with base model to see if the embedding layers weight has changed or not. It didn't change. However, the layers weight has changed

**Reply by BenjaminBossan** [reply]
Oh, so the 0s are not the weights, but the difference. How did you calculate the difference?

**Reply by RADxEEn** [reply]
So I have model structure suppose one is base_model and the other one is merged_model which is being merged with base and adapter model after training.

`base_model.model.model.lm_head.original_moduleSS.weight-.base_model.model.model.lm_head.modules_to_save.default.weight`

which should be give some difference between the weight. But as giving same weight, so after taking difference, it is giving 0.

**Reply by BenjaminBossan** [reply]
Sorry, I still don't quite understand how you calculated that, maybe you could provide more code?

> However, the layers weight has changed

So is it possible that the update worked correctly?

**Reply by RADxEEn** [reply]
Suppose the structure is like below for both code (base model and merged_model_after_train).
So for base model we can get the embedding weight by using
```
MistralForCausalLM(
  (model): MistralModel(
    (embed_tokens): Embedding(67840, 4096)
    (layers): ModuleList(
      (0-31): 32 x MistralDecoderLayer(
        (self_attn): MistralAttention(
          (q_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)
          (k_proj): Linear4bit(in_features=4096, out_features=1024, bias=False)
          (v_proj): Linear4bit(in_features=4096, out_features=1024, bias=False)
          (o_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)
          (rotary_emb): MistralRotaryEmbedding()
        )
        (mlp): MistralMLP(
          (gate_proj): Linear4bit(in_features=4096, out_features=14336, bias=False)
          (up_proj): Linear4bit(in_features=4096, out_features=14336, bias=False)
          (down_proj): Linear4bit(in_features=14336, out_features=4096, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): MistralRMSNorm()
        (post_attention_layernorm): MistralRMSNorm()
      )
    )
    (norm): MistralRMSNorm()
  )
  (lm_head): Linear(in_features=4096, out_features=67840, bias=False)
)
```

**model.embed_tokens.weight** which will generate the tensor of the weight for embedding module.
same for the merged model.

Then if I take the difference between these two tensors, it should not be 0, as it should be updating, like the self_attention layer module.

If I calculate the difference for self_attention layer between two model, i have found that the difference is giving me non zero tensor, which is not the case for embedding layer weight, even though I use modules_to_save. 


So, what I am thinking is that, I missing something or is the merging the adapter with the base model is not doing properly?

Here I am giving the merging code

```
from peft import PeftModel
adapter_path="/content/bn_adapter"
merged_model_f_n=PeftModel.from_pretrained(base_model,adapter_path)
merged_model_f=merged_model_f_n.merge_and_unload()
```

**Reply by BenjaminBossan** [reply]
So one thing I see is that you're using 4bit quantization. Is the quantization also applied to the embedding layer? Since this should make it so that the embedding layer cannot be trained.

**Reply by RADxEEn** [reply]
Yes the quantization also applied in the embedding layer. Anyway, the issue
has been resolved. The problem is with the colab. As they have a tendency
to optimize the vram, although i am calling with different variable,somehow
they use the cache to store the data. For this, if I call the models, I
don’t know but it automatically updates the base model similar to the
merged model. For this, i am getting zero. Now in kaggle, the problem has
been solved. Thank you.
